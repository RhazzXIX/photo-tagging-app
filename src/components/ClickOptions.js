const ClickOptions = (props) => {
  const { selections, position } = props;

  let posY = 5;
  let posX = 5;
  if (position.targetPosY) posY += position.targetPosY;
  if (position.targetPosX) posX += position.targetPosX;
  if (posY >= 910) posY -= 100;
  if (posX >= 1680) posX -= 174;

  return (
    <ul
      style={{
        position: "absolute",
        top: posY,
        left: posX,
      }}
    >
      {selections.map((character) => {
        if (character.isFound) return null;
        return (
          <li key={character.id}>
            <button onClick={() => character.fn()}>{character.name}</button>
          </li>
        );
      })}
    </ul>
  );
};

export default ClickOptions;
