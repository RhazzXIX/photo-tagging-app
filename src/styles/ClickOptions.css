ul#charBtns {
  display: flex;
  flex-direction: column;
  width: 160px;
  height: 160px;
  justify-content: space-evenly;
  align-items: center;
  border-radius: 8px;
  background-color: rgba(255, 255, 255, 0.4);
}

ul#charBtns button {
  width: 146px;
  height: 50px;
  border-radius: 12px;
  font-size: 1.2rem;
  display: flex;
  justify-content: center;
  align-items: center;
}
