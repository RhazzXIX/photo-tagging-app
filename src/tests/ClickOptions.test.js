import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { screen, render } from "@testing-library/react";
import ClickOptions from "../components/ClickOptions";

const luffy = {
  name: "Luffy",
  fn(e) {
    this.isFound = true;
  },
  id: 1,
  isFound: false,
};

const selections = [
  {
    name: "Heli",
    fn: jest.fn(),
    id: 2,
  },
  {
    name: "Poli",
    fn: jest.fn(),
    id: 3,
  },
];

const user = userEvent.setup();

describe("ClickOptions component", () => {
  beforeEach(() => {
    luffy.isFound = false;
    selections.push(luffy);
  });

  it("Renders on screen", () => {
    const { container } = render(<ClickOptions selections={selections} />);
    expect(container).toMatchSnapshot();
  });

  it("The button activates the appropriate functions of each selection", async () => {
    render(<ClickOptions selections={selections} />);
    const buttons = screen.getAllByRole("button");

    await user.click(buttons[2]);

    expect(selections[2].isFound).toBeTruthy();
    expect(selections[1].fn).not.toHaveBeenCalled();

    await user.click(buttons[0]);
    await user.click(buttons[1]);

    expect(selections[0].fn).toHaveBeenCalled();
    expect(selections[1].fn).toHaveBeenCalled();
  });

  it("Doesn't render the button of a found character", async () => {
    const { rerender } = render(<ClickOptions selections={selections} />);
    const luffyBtn = screen.getByRole("button", { name: "Luffy" });
    let buttons = screen.getAllByRole("button");

    expect(buttons).toHaveLength(3);

    await user.click(luffyBtn);
    rerender(<ClickOptions selections={selections} />);
    buttons = screen.getAllByRole("button");

    expect(
      screen.queryByRole("button", { name: "Luffy" })
    ).not.toBeInTheDocument();
    expect(buttons).toHaveLength(2);
  });
  afterEach(() => selections.pop());
});
