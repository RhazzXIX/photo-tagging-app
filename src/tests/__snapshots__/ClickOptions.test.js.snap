// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickOptions component Renders on screen 1`] = `
<div>
  <ul
    style="position: absolute; top: 5px; left: 5px;"
  >
    <li>
      <button>
        Heli
      </button>
    </li>
    <li>
      <button>
        Poli
      </button>
    </li>
    <li>
      <button>
        Luffy
      </button>
    </li>
  </ul>
</div>
`;
